app/services/usps_in_person_proofing/proofer.rb

```diff
@@ -102,7 +103,14 @@ def request_enrollment_code(unique_id)
     # @return [String] the token
     def retrieve_token!
       body = request_token
-      expires_at = Time.zone.now + body['expires_in']
+      # Refresh our token early so that it won't expire while a request is in-flight. We expect 15m
+      # expirys for tokens but are careful not to trim the expiry by too much, just in case
+      expires_in = body['expires_in'].seconds
+      if expires_in - AUTH_TOKEN_PREEMPTIVE_EXPIRY_MINUTES > 0
+        expires_in -= AUTH_TOKEN_PREEMPTIVE_EXPIRY_MINUTES
+      end
+
+      expires_at = Time.zone.now + expires_in
       token = "#{body['token_type']} #{body['access_token']}"
       Rails.cache.write(AUTH_TOKEN_CACHE_KEY, token, expires_at: expires_at)
```

Comment:
I think we may want to use `expires_in` rather than `expires_at`?

https://github.com/rails/rails/pull/45047 adds Redis/Memcached support for `expires_at`, but it hasn't been released yet (I assume it will be in 7.1).
```
[1] pry(main)> Rails.cache.write('1', '2', expires_in: 10.seconds)
=> "OK"
[2] pry(main)> REDIS_POOL.with { |x| x.ttl('1') }
=> 8
[3] pry(main)> Rails.cache.write('1', '2', expires_at: 10.second.from_now)
=> "OK"
[4] pry(main)> REDIS_POOL.with { |x| x.ttl('1') }

Reply:
Last I checked neither `expires_in` or `expires_at` were respected by our redis configuration when using `Rails.cache.write`. You can see observed behavior in this [slack thread](https://gsa-tts.slack.com/archives/C03FA4VBN76/p1667592731100459?thread_ts=1667514804.376859&cid=C03FA4VBN76). That's why we set the redis ttl just below this, [here](https://github.com/18F/identity-idp/pull/8035/files#diff-21dd6f37ce8c3f0ded422c037611d8775e409008e522790a272e2b3b4b426091R116-R120)

Reply:
I tested it out in the Joy environment and it does appear to be working now. I suppose it was fixed or began to be supported in a recent upgrade we did. I'll go ahead and update this PR to remove the redis hack and test it more thoroughly in the Joy sandbox

Reply:
To be clear I plan to:
1. Update #retrieve_token! to use `expires_in` instead of `expires_at`
2. Remove the hacky redis ttl workaround
3. Update specs to not expect the hacky redis ttl workaround
4. Deploy to Joy sandbox and test that caching works as expected

Reply:
I was able to confirm this change works as expected. For example, here's a demo of it deployed to the Joy sandbox:

```
irb(main):001:1* def get_token # defining a helper method to print the time and the current token
irb(main):002:1*   puts Time.zone.now
irb(main):003:1*   puts UspsInPersonProofing::Proofer.new.token
irb(main):004:0> end
=> :get_token
irb(main):005:0> get_token # notice here that the cache is cold and so an API request is made to retrieve a token
2023-03-22 19:02:39 UTC
{"http_method":"POST","host":"<redacted>","path":"/oauth/authenticate","duration_seconds":0.816173831,"status":200,"service":"usps_token","name":"request_metric.faraday"}
Bearer uqL<redacted>
=> nil
irb(main):055:0> get_token # notice here that the cache is hot so no API request is made, and the same token is printed
2023-03-22 19:16:38 UTC
Bearer uqL<redacted>
=> nil
irb(main):056:0> get_token
2023-03-22 19:16:39 UTC
Bearer uqL<redacted>
=> nil
irb(main):057:0> get_token # notice here, just over 14 minutes from the last API request, the cache is now cold again and so an API request is made to retrieve a token
2023-03-22 19:16:41 UTC
{"http_method":"POST","host":"<redacted>","path":"/oauth/authenticate","duration_seconds":0.653552645,"status":200,"service":"usps_token","name":"request_metric.faraday"}
Bearer kTa<redacted>
=> nil
```

I did the same procedure locally with the same results